### ISSUE TITLE
Per allele field encoding

### ISSUE THREAD

**Issue by d-cameron** [samtools/hts-specs#133]
Writing some test cases for symbolic allele parsing, I've encountered some problems with the SV specifications when dealing with multiple SV alt alleles since only some fields support multiple alleles:
- Spec-defined field counts are inconsistent. SVLEN, HOMLEN, HOMSEQ allows any number of values, but NOVEL, IMPRECISE, END, SVTYPE all require only a single value.
- CIPOS and CIEND, CILEN require exactly 2 values,

There are also more general problems with combining allele-specific fields.
- There is no spec-supported way to specify allele-specific lists. MATEID is a per allele-list so if there a 3 alt alleles with 1, 2 and 3 mates respectively, encoding this as 6 IDs in and/or EVENT fields loses information since there is no way to recover which alt allele was associated with which mate (looking for reciprocal MATEIDs 

There are two solutions to this:
1.  Update specs to resolve these issues by a) Updating ##INFO<Number= with a record-per-allele or N-records-per-allele (for CIPOS-style fields); b) Updated SV fields to match and c) add variable-count-per-allele and associated list-of-list syntax
2. Update the specifications to not allow symbolic alleles in records with more than one alt allele

I strongly favour 2 for the following reasons:
* 1 is a large change to the specifications and, given the poor tooling support for SVs currently, would increase this barrier further
* No tools (that I'm aware of) actually write multiple symbolic alt alleles in a single record
* Many of the advantages of allowing multiple alt allele in a single record do not apply to SVs

**Comment by d-cameron** [comment]
1.a) is already in VCF 4.2 but the list of lists is the key issue that is still problematic in the 4.3 draft

**Comment by cyenyxe** [comment]
For the sake of completion, these are all the problematic SV fields described in v4.3 of the specification along with their Number and Type. There is quite a lot of them.

### Section 3
| ID | Number | Type |
|---|---|---|
| IMPRECISE | 0-1 | Flag |
| NOVEL | 0-1 | Flag |
| END | 1 | Integer |
| SVTYPE | 1 | String |
| CIPOS | 2 | Integer |
| CIEND | 2 | Integer |
| MEINFO | 4 | String |
| METRANS | 4 | String |
| DGVID | 1 | String |
| DBVARID | 1 | String |
| PARID | 1 | String |
| EVENT | 1 | String |
| CILEN | 2 | Integer |
| DP | 1 | Integer |
| CN | 1 | Integer |
| CICN | 2 | Integer |

As a side comment, wouldn't DGVID and DBVARID fit better in the ID column?

### Section 4
| ID | Number | Type |
|---|---|---|
| CN | 1 | Integer |
| CNQ | 1 | Float |
| NQ | 1 | Float |
| HAP | 1 | Integer |
| AHAP | 1 | Integer |

**Comment by cyenyxe** [comment]
Option 1 would be more aligned with the rest of the specification, but the list-of-list syntax could make definitions even more complex than they already are. Option 2 sounds definitely simpler, but more people would need to confirm it wouldn't be a big issue to drop this partial support for multi-allelic SV variants.

I know it is being a while, but could you please provide a list of tools that you are sure do not write multiple symbolic alleles in a single record? It would really help to have a reference of them.

**Comment by d-cameron** [comment]
The following SV callers do not write in VCF format:
```
breakdancer
gasv
gasvpro/gasvpro
variantionhunter/commonlaw
crest
hydra/hydra-multi
prism
SVMerge
SOAPsv
SRiC
SVseq/SVseq2
ClipCrop
SVM2
```
The following SV callers can write in VCF but I have not found any output in which multiple symbolic alleles are written to the same record:
```
clever
cortex
delly
gridss
lumpy
manta
pindel
socrates
tigra
```

The only case I've encountered where multiple symbolic alt alleles are actually written to a VCF is from running `bcftools merge` to merge the delly per-event type outputs.

TLDR: I don't think this actually happens in the wild yet.

**Comment by magicDGS** [comment]
Will the restriction be applied to **all** symbolic alleles? Symbolic alleles could be SVs, but also any other user-defined ones. For example, IUPAC codes for ambiguous bases may be useful in the same VCF line.

I personally do not like to have a restriction in all the symbolic alleles just because there are problematic with SVs. I rather prefer to specify that the SV-specific fields are constraint to contain information of just one event, and left the rest as-is for other symbolic alleles.

This is just my opinion, which is biased because I do not work with SVs.

**Comment by d-cameron** [comment]
>I personally do not like to have a restriction in all the symbolic alleles just because there are problematic with SVs

I hadn't noticed that IUPAC ambiguity codes are now allowed as symbolic alleles. I'd be happy for it only to apply to SV symbolic/breakpoint/breakend alleles.

**Comment by thefferon** [comment]
I agree with the suggestion that the sorts of restrictions being discussed only apply to SV alleles.

For the sake of completeness of discussion, what is the primary argument(s) for **allowing** multiple symbolic alleles per VCF line (and the resulting complications that necessarily follow)? If it mainly involves keeping VCF "tidy" by only having one instance of any given CHROM-POS doublet in a VCF file, then perhaps other concerns outweigh it. If, on the other hand, there are dependencies or implications in callers and analyses (possibly including any that may arise when SNV and SV data are integrated within a VCF), then that may be another story.

**Comment by d-cameron** [comment]
>For the sake of completeness of discussion, what is the primary argument(s) for allowing multiple symbolic alleles per VCF line (and the resulting complications that necessarily follow)?

I'm not aware of one other than than the VCF format supported multiple ALT alleles before SV support was added. I've never encountered an SV caller that actually writes two SVs to the same record. SNV/indel callers, yes, but they use the standard ALT allele notation, not the SV notation.

**Comment by cyenyxe** [comment]
The reason stated by @d-cameron is the only one I can think of. Do you know of any other @pd3 ?

If we make the rule absolutely clear and unambiguous in the spec, I am fine with supporting multiple SNP/INDEL but only one SV per record.

**Comment by pd3** [comment]
There certainly are legitimate cases where multiple symbolic alleles are desirable, such as copy number variations (see the `<CNx>` alleles in the 1000 Genomes Project VCFs).  Therefore I agree with @magicDGS that such restriction should not apply to all symbolic alleles in general. From the design perspective, it feels ad-hoc to allow multiallelic records for one variation type but not for others, so I am leaning slightly towards making this consistent.
However, I am not working with SVs myself much and therefore I don't have a strong opinion on this.

**Comment by magicDGS** [comment]
What's about option 1 in the case of multiple symbolic alleles to make it consistent with the rest of the specification (as pointed by @cyenyxe), but recommending to encode only a single SV per 
 recommendation for SVs to have only one ALT allele?

Like that, all symbolic alleles will have the same specifications and in the case of SVs is up to the developers to use the recommended encoding (which will make their life simpler) or stick to the multi-allelic representation.

**Comment by d-cameron** [comment]
> in the case of SVs is up to the developers to use the recommended encoding (which will make their life simpler) or stick to the multi-allelic representation.

The issue is that the specifications themselves don't support many multi-allelic symbolic SV representation due to the definitions of various SV info fields. There is no way to represent both an insertion and a deletion using SV symbolic alleles in a single record in a spec-compliant manner because SVTYPE is required but must be a single value.

Option 1 is a major change to the specifications as it requires a new INFO field list-of-lists encoding format. It's a large burden on VCF parsers for something that's likely (almost) never going to be used. It's unlikely to be backwards comparable as we'd probably have to reserve an additional character as a list separator. A half-way option would be to disallow multi-symbolic records only for variants in which the list-of-lists ambiguities arise.

**Comment by magicDGS** [comment]
I understand that it is a disruptive change, but I can see more usages of a new list-of-lists encoding: for instance, effect of variants per-transcript. In addition, it solves the issue with multi-allelic symbolic SV.

Anyway, I do not have any strong opinion on this because I haven't started yet to deal with SV encoding (although I'm planning to do it).
